File: 210.course-schedule-ii.py
#
# @lc app=leetcode id=210 lang=python3
#
# [210] Course Schedule II
#

# @lc code=start
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        graph = [[] for _ in range(numCourses)]
        for edge in prerequisites:
            graph[edge[0]].append(edge[1])

        state = [0 for _ in range(numCourses)]
        ans = []
        
        def dfs(node: int) -> bool:
            state[node] = 1
            for t in graph[node]:
                if state[t] == 1:
                    return False
                if state[t] == 0 and not dfs(t):
                    return False
            state[node] = 2
            ans.append(node)
            return True
        
        for i in range(numCourses):
            if state[i] == 0 and not dfs(i):
                return []
        
        return ans
# @lc code=end

File: 210.course-schedule-ii.cpp
/*
 * @lc app=leetcode id=210 lang=cpp
 *
 * [210] Course Schedule II
 */

// @lc code=start
class Solution {
 public:
  enum Color { white, grey, black };
  vector<int> findOrder(int numCourses, vector<vector<int>> &prerequisites) {
    vector<vector<int>> graph(numCourses);
    vector<Color> record(numCourses, white);
    vector<int> ans;

    for (auto &p : prerequisites) {
      graph[p[1]].push_back(p[0]);
    }

    for (int i = 0; i < numCourses; ++i) {
      if (record[i] == white and !dfs(graph, i, record, ans)) {
        return {};
      }
    }

    reverse(ans.begin(), ans.end());
    return ans;
  }

  bool dfs(vector<vector<int>> &graph, int i, vector<Color> &record,
           vector<int> &ans) {
    record[i] = grey;
    for (int j : graph[i]) {
      if (record[j] == grey) {
        return false;
      } else if (record[j] == white and !dfs(graph, j, record, ans)) {
        return false;
      }
    }
    record[i] = black;
    ans.push_back(i);
    return true;
  }
};
// @lc code=end
